### PR TITLE
Implementation of Standard k-ε Turbulence Model within ESPH

### PR DESCRIPTION
**Pull request by Yashmandaokar**
This pull request introduces the implementation of the standard k-ε turbulence model with the standard wall function in the ESPH framework. The implementation is verified against a 2D turbulent channel flow case at a Reynolds number of 20,000. Additionally, the PR ensures naming consistency by renaming a test case.

1. Standard k-ε Turbulence Model Implementation:
The standard k-ε transport equations are implemented along with a standard wall function for turbulence modeling.

Files: turbulence_model.h, turbulence_model.cpp, and turbulence_model.hpp:
- Definition and implementation of k-ε transport equations.
- Implementation of the standard wall function, wall-adjacent cell identification.

2. HLLC Riemann Solver for Turbulence Variables:
Extends the Eulerian HLLC Riemann solver to handle the advection terms for turbulence variables (k and ε).

Files: extended_eulerian_riemann_solver.h and extended_eulerian_riemann_solver.cpp:
- Extends the HLLC Riemann solver to compute interface states for TKE and dissipation variables.

3. Turbulence Forces for RANS Momentum Equation:
Implements the turbulence forces necessary for the RANS momentum equation: 

- TKE Gradient Force.
- Turbulent Viscous Force.

Files: rans_turbulence_dynamics.h and rans_turbulence_dynamics.hpp:
- Implements turbulence force calculations for incompressible/weakly compressible flows.

4. Test Case: test_2d_FVM_turbulent_channelflow:
The test_2d_FVM_turbulent_channelflow test case is developed to verify the turbulence model implementation:

Verification Details:
- Reynolds number: 20,000.
- Verification against Fluent results.

5. Updates to the test_3d_incompressible_channel_flow case:

- Naming Consistency: The test case test_3d_incompressible_channel_flow is renamed to test_3d_FVM_incompressible_channel_flow to align with the naming conventions for FVM test cases.
- Rearrangement: To avoid double registration of the variables, some part of the code is rearranged
- Initialization: Mass and momentum variables initializaton is added, which results in reduction in simulation time.

copilot generated summary:

This pull request includes several changes to improve the fluid dynamics simulations and refactor the test files for better organization and functionality. The changes focus on adding turbulence models, refactoring test files, and minor code cleanups.

### Improvements to fluid dynamics simulations:

* [`src/shared/particle_dynamics/fluid_dynamics/eulerian_fluid_dynamics/eulerian_fluid_integration.hpp`](diffhunk://#diff-f5b3300ed3ae5f241eec5d02985a186033830decbc4b1cb428e0723c6ec217d5R41): Added turbulence model calculations for `EulerianIntegration1stHalf` and `EulerianIntegration2ndHalf` interactions. [[3]](diffhunk://#diff-f5b3300ed3ae5f241eec5d02985a186033830decbc4b1cb428e0723c6ec217d5R41) [[2]](diffhunk://#diff-f5b3300ed3ae5f241eec5d02985a186033830decbc4b1cb428e0723c6ec217d5R109)
* [`tests/extra_source_and_tests/extra_src/for_2D_build/turbulence_model.cpp`](diffhunk://#diff-cb73d32fcd44e69bf5337e1e510c6721f2243dcb89f085d39cb351f30c628c3eR1-R130): Introduced new turbulence model implementations, including `BaseTurbulence`, `WallAdjacentCells`, and `StdWallFunctionFVM` classes.

### Refactoring and renaming test files:

* [`tests/3d_examples/test_3d_FVM_incompressible_channel_flow/test_3d_FVM_incompressible_channel_flow.cpp`](diffhunk://#diff-7f9264e72aadb90bd8e4ee397fa6502de564a1ef9f08881a75d2afe9e6c3cf08L2-R6): Renamed the file and updated references to reflect the new name for consistency. [[1]](diffhunk://#diff-7f9264e72aadb90bd8e4ee397fa6502de564a1ef9f08881a75d2afe9e6c3cf08L2-R6) [[2]](diffhunk://#diff-7f9264e72aadb90bd8e4ee397fa6502de564a1ef9f08881a75d2afe9e6c3cf08L38-R42) [[3]](diffhunk://#diff-7f9264e72aadb90bd8e4ee397fa6502de564a1ef9f08881a75d2afe9e6c3cf08L60-R62)
* [`tests/3d_examples/test_3d_FVM_incompressible_channel_flow/test_3d_FVM_incompressible_channel_flow.h`](diffhunk://#diff-84f55aa9a799a2f7616ee2f915fc7883da6266b9071ad7544c6f20e2dd3c15f4L2-R8): Renamed the file and updated references to reflect the new name for consistency. [[1]](diffhunk://#diff-84f55aa9a799a2f7616ee2f915fc7883da6266b9071ad7544c6f20e2dd3c15f4L2-R8) [[2]](diffhunk://#diff-84f55aa9a799a2f7616ee2f915fc7883da6266b9071ad7544c6f20e2dd3c15f4L27) [[3]](diffhunk://#diff-84f55aa9a799a2f7616ee2f915fc7883da6266b9071ad7544c6f20e2dd3c15f4L60-R62) [[4]](diffhunk://#diff-84f55aa9a799a2f7616ee2f915fc7883da6266b9071ad7544c6f20e2dd3c15f4R73-R80)

### Minor code cleanups:

* [`tests/2d_examples/test_2d_FVM_flow_around_cylinder/common_weakly_compressible_FVM_classes.h`](diffhunk://#diff-e7781f81edd243ff07f1db891eca6dcd634340c98bf590ca814499c74f56f3eeL142): Removed unnecessary blank lines for better code readability.

### CODE REVIEW

**Comment by Yashmandaokar** [comment]
@FengWang3119  Could you please review this pull request when you have a chance? Thank you!

File: tests/3d_examples/test_3d_FVM_incompressible_channel_flow/test_3d_FVM_incompressible_channel_flow.cpp
Comment (Yashmandaokar) [inline tests/3d_examples/test_3d_FVM_incompressible_channel_flow/test_3d_FVM_incompressible_channel_flow.cpp:1]: 1. Case file renamed for consistency with the 2d FVM cases.
2. Initial condition postion rearanged to avoid double registration of the variables.
3. mass and momentum initializaton helps faster convergence hence the total simulation time is reduced

File: tests/3d_examples/test_3d_FVM_incompressible_channel_flow/test_3d_FVM_incompressible_channel_flow.h
Comment (Yashmandaokar) [inline tests/3d_examples/test_3d_FVM_incompressible_channel_flow/test_3d_FVM_incompressible_channel_flow.h:1]: 1. Renamed for consistency with 2d FVM
2. Remove redundant variable
3. Initializing mass and momentum variable

**Comment by Yashmandaokar** [comment]
Hello @FengWang3119 and  @yyc84 

I’m facing issues with the CI tests:
test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?

 test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?

Your guidance would be greatly appreciated!

**Comment by FengWang3119** [comment]
> Hello @FengWang3119 and @yyc84
> 
> I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> 
> test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> 
> Your guidance would be greatly appreciated!

Is is possible to shorten the channel length?

**Comment by Yashmandaokar** [comment]
> > Hello @FengWang3119 and @yyc84
> > I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> > test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> > Your guidance would be greatly appreciated!
> 
> Is is possible to shorten the channel length?

YES, it is possible but will require lot of rework. I will first try reducing the simulation time and increasing the CFL number from 0.6 to 1.0. Thanks for your input. 
Do you have any clue why the second case: test_2d_mixed_poiseuille_flow is falling only on windows?

**Comment by Yashmandaokar** [comment]
Dear Professor @Xiangyu-Hu 

The test case test_2d_FVM_turbulent_channelflow is failing because of timeout. The channel is long in order to achieve fully developed turbulence (also becasue:  periodic boundary conditions are unavailable in the FVM framework), hence the simulation time is also long. 

Could you please advice me, how to solve the timeout issue?

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        MultiPolygon water_block(createWaterBlockShape());
+        add<MultiPolygonShape>(water_block, "WaterBlock");
+    }
+};
+//----------------------------------------------------------------------
+//	Initialization
+//----------------------------------------------------------------------
+
+class TCFInitialCondition
```
Comment (FengWang3119) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:65]: For the class name, please use full name, a long name would not cause trouble but make your code more clear.
Currently, I am confused with this TCF

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+//	Material properties of the fluid.
+//----------------------------------------------------------------------
+Real rho0_f = 1.0;
+Real U_f = 1.0;
+Real rey_bulk = 20000.0;
+Real mu_f = (rho0_f * U_f * DH * 0.5) / rey_bulk;       /**< Dynamic Viscosity. */
+Real c_f = 10.0 * U_f;                               /**< Reference sound speed. */
+Real C_mu = 0.09;
+
+Real I = 0.05;
```
Comment (FengWang3119) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:33]: explanation please, turbulent intensity?  if possible, avoid using the single character which may cause conflicts with other varaibles.

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/common_weakly_compressible_FVM_classes.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
+
+#include "common_weakly_compressible_FVM_classes.h"
```
Comment (FengWang3119) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/common_weakly_compressible_FVM_classes.cpp:2]: Just for curious, why we have this file that not very relevant with the turbulent module?

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h (hunk 1 above)
Comment (Yashmandaokar) [reply tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:65]: Sure, I will change it to full name

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h (hunk 2 above)
Comment (Yashmandaokar) [reply tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:33]: Yes I here stands for turbulent intensity. I will also use a full name here.

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/common_weakly_compressible_FVM_classes.cpp (hunk 3 above)
Comment (Yashmandaokar) [reply tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/common_weakly_compressible_FVM_classes.cpp:2]: In FVM it is need to calculate WCAcousticTimeStepSizeInFVM.

**Comment by Xiangyu-Hu** [comment]
@Yashmandaokar It seems that there is some incompatibility to the macOS system. You can try to compile the code using clang compiler in Linux to have a check first. If it also gives the same problem, you can debug the code and find the bug.

**Comment by Xiangyu-Hu** [comment]
> Hello @FengWang3119 and @yyc84
> 
> I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> 
> test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> 
> Your guidance would be greatly appreciated!

I am not expecting that a 2d test case requires more than 1000s. How many cells have you used in this case?

**Comment by Yashmandaokar** [comment]
> > Hello @FengWang3119 and @yyc84
> > I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> > test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> > Your guidance would be greatly appreciated!
> 
> I am not expecting that a 2d test case requires more than 1000s. How many cells have you used in this case?

The case has 55322 cells. It is a long channel length 120 m heigth 2 m.  The channel is long in order to achieve fully developed turbulence (also becasue: periodic boundary conditions are unavailable in the FVM framework)

**Comment by Xiangyu-Hu** [comment]
> > > Hello @FengWang3119 and @yyc84
> > > I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> > > test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> > > Your guidance would be greatly appreciated!
> > 
> > 
> > I am not expecting that a 2d test case requires more than 1000s. How many cells have you used in this case?
> 
> The case has 55322 cells. It is a long channel length 120 m heigth 2 m. The channel is long in order to achieve fully developed turbulence (also becasue: periodic boundary conditions are unavailable in the FVM framework). Also the simulation time is 140 sec in order to reach steady state. Should I make the mesh coarse?

@Yashmandaokar It is also not meaning full to use 60 time of the height as the length. 
@FengWang3119 Hi, Feng, do we really need such long channel?

**Comment by FengWang3119** [comment]
> > > > Hello @FengWang3119 and @yyc84
> > > > I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> > > > test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> > > > Your guidance would be greatly appreciated!
> > > 
> > > 
> > > I am not expecting that a 2d test case requires more than 1000s. How many cells have you used in this case?
> > 
> > 
> > The case has 55322 cells. It is a long channel length 120 m heigth 2 m. The channel is long in order to achieve fully developed turbulence (also becasue: periodic boundary conditions are unavailable in the FVM framework). Also the simulation time is 140 sec in order to reach steady state. Should I make the mesh coarse?
> 
> @Yashmandaokar It is also not meaning full to use 60 time of the height as the length. @FengWang3119 Hi, Feng, do we really need such long channel?

For the freestream inlet conditions, the length seems tobe neccessary for the turbulent kinetic energy to achieve fully-developed acoording to the FVM result. 
But if impose the inlet data from the fully-developed result, 15 time is OK. 
I remembered I told Yash before, but it seems there is difficulty for the implementation.

**Comment by Yashmandaokar** [comment]
> > > > > Hello @FengWang3119 and @yyc84
> > > > > I’m facing issues with the CI tests: test_2d_FVM_turbulent_channelflow: All tests fail due to a timeout error. The channel is long, requiring a longer duration to complete. I know I am not allowed to modify the ci.yml file. What other options do I have to pass the test?
> > > > > test_2d_mixed_poiseuille_flow: Fails only on Windows. It works fine on other platforms. Any idea why?
> > > > > Your guidance would be greatly appreciated!
> > > > 
> > > > 
> > > > I am not expecting that a 2d test case requires more than 1000s. How many cells have you used in this case?
> > > 
> > > 
> > > The case has 55322 cells. It is a long channel length 120 m heigth 2 m. The channel is long in order to achieve fully developed turbulence (also becasue: periodic boundary conditions are unavailable in the FVM framework). Also the simulation time is 140 sec in order to reach steady state. Should I make the mesh coarse?
> > 
> > 
> > @Yashmandaokar It is also not meaning full to use 60 time of the height as the length. @FengWang3119 Hi, Feng, do we really need such long channel?
> 
> For the freestream inlet conditions, the length seems tobe neccessary for the turbulent kinetic energy to achieve fully-developed acoording to the FVM result. But if impose the inlet data from the fully-developed result, 15 time is OK. I remembered I told Yash before, but it seems there is difficulty for the implementation.

At the early stage, my focus was on reducing complexity, as the code was already encountering several other issues. I did not anticipate that a 15-minute simulation would pose a problem—apologies for that oversight. However, I can now look into shortening the channel length (to 15 times of height) and implementing a fully developed velocity profile at the inlet, if that would be an acceptable approach. Should I proceed with this approch?

**Comment by FengWang3119** [comment]
> At the early stage, my focus was on reducing complexity, as the code was already encountering several other issues. I did not anticipate that a 15-minute simulation would pose a problem—apologies for that oversight. However, I can now look into shortening the channel length (to 15 times of height) and implementing a fully developed velocity profile at the inlet, if that would be an acceptable approach. Should I proceed with this approch @Xiangyu-Hu and @FengWang3119 ?

Sure, it seems to be the only way to resolve this issue.

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/data/Channel_Mesh_short.msh
Comment (Yashmandaokar) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/data/Channel_Mesh_short.msh:1]: Mesh file for new mesh (short channel)

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    void applyNonSlipWallBoundary(size_t ghost_index, size_t index_i) override
+    {
+        vel_[ghost_index] = -vel_[index_i];
+        p_[ghost_index] = p_[index_i];
+        rho_[ghost_index] = rho_[index_i];
+        K_[ghost_index] = K_[index_i];
+        Eps_[ghost_index] = Eps_[index_i];
+        mu_t_[ghost_index] = mu_t_[index_i];
+    }
+    //Fully Developed Turbulence Profiles
```
Comment (Yashmandaokar) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:122]: Importing fully developed turbulence profiles

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                                                          0.0045933602, 0.0054699243, 0.007050876, 0.0061376584};
+
+    std::vector<Real> turbulent_dissipation_rate_profile = {0.0056344033, 0.0023513641, 0.0011376451, 0.00069459993,
+                                                            0.00046991525, 0.00032542294, 0.00022683406, 0.0001672248,
+                                                            0.00013027232, 0.00011116108, 0.00011085657, 0.00013127667,
+                                                            0.00017384913, 0.00023951771, 0.00032356411, 0.00043955375,
+                                                            0.00067517039, 0.0011331175, 0.0023402071, 0.0056611444};
+    // Function to find the nearest value based on the y-coordinate
+    Real nearestValue(Real y, const std::vector<Real> &y_values, const std::vector<Real> &data_values)
+    {
```
Comment (Yashmandaokar) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:148]: Apply constant interpolation on the turbulent profiles

File: tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            {
+                closest_index = i;
+                min_distance = distance;
+            }
+        }
+        return data_values[closest_index];
+    }
+
+        void applyVelocityInletFlow(size_t ghost_index, size_t index_i) override
+        {
```
Comment (Yashmandaokar) [inline tests/extra_source_and_tests/test_2d_FVM_turbulent_channelflow/test_2d_FVM_turbulent_channelflow.h:165]: Imposing the fully developed profiles on the inlet bc
